### PR TITLE
Introduce double buffered HULA reader

### PR DESCRIPTION
**Pull request by h3ndrk**
## Introduced Changes

This PR introduces Unix socket draining when reading from the HULA socket. This is needed because our HULK process connects very early to the Unix socket and HULA/LoLA immediately starts sending state messages to HULK which fills our read buffer. After the connection, we are further initializing the HardwareInterface, e.g. doing camera initialization/resetting. This may take a long time, filling up around 1 second of historic state messages in HULKs read buffer. Once the HardwareInterface has been initialized, the cyclers begin to run and the control cycler will rush to drain the accumulated messages, because the sensor data receiver has a lot of data available to feed the cycle. This results in a lot of messages being sent to HULA/LoLA containing historic (1 second old) control messages. But since HULA/LoLA only updates its shared memory and does not communicate directly with the hardware, this is usually not noticed and it will converge quickly to nearly no latency and drained buffers. This bug was discovered because the new HULA will communicate synchronously with the hardware thus needing this fix.

Changes in HULA are not needed. No worries, there is a lot of test code which accumulates a lot of added lines. :wink: 

## ToDo / Known Issues

None

## Ideas for Next Iterations (Not This PR)

None

## How to Test

The double buffering should have nearly 100% test coverage, but please test on a real NAO if everything works as intended. The chest button LED blinking frequency should be exactly one second immediately on startup which is a good indicator that everything works as intended.

### CODE REVIEW

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 1]
```diff
@@ -0,0 +1,411 @@
+use std::{
+    io::{self, ErrorKind, Read},
+    mem::size_of,
+    os::unix::{io::AsRawFd, prelude::RawFd},
+};
+
+use super::hula::StateStorage;
+
+pub struct DoubleBufferedReader {
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:9]: What do you think of this with some generics in an extra crate? Or is it overkill?

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+            buffers: [[0; size_of::<StateStorage>()]; 2],
+            active_buffer_index: Default::default(),
+            incomplete_offset: Default::default(),
+        }
+    }
+}
+
+impl DoubleBufferedReader {
+    fn swap_buffers(&mut self) {
+        self.incomplete_offset = 0;
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:27]: `number_of_incomplete_bytes` ?

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        self.active_buffer_index = (self.active_buffer_index + 1) % 2;
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:28]: `2` -> `NUMBER_OF_BUFFERS` ?

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    }
+
+    pub fn read(
+        &mut self,
+        reader: &mut (impl AsRawFd + Read),
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:33]: An idea:
Is it possible to split the `UnixStream` into an reader and writer part?
I want to reach a point where the reader can be consumed/owned by the `DoubleBufferedReader`.
This means, moving a hypothetical reading part to `Self` and drop the argument here.

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+        mut poll_reader: impl FnMut(RawFd) -> io::Result<()>,
+    ) -> io::Result<()> {
+        let mut buffers_swapped = false;
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:36]: Naming things: This boolean, is used to express whether a full message was received since the last read, correct? Maybe something like `is_buffer_complete`?

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+        loop {
+            match reader.read(&mut self.buffers[self.active_buffer_index][self.incomplete_offset..])
+            {
+                Ok(number_of_read_bytes) => {
+                    self.incomplete_offset += number_of_read_bytes;
+                    assert!(self.incomplete_offset <= self.buffers[self.active_buffer_index].len());
+                    if self.incomplete_offset == self.buffers[self.active_buffer_index].len() {
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:43]:
```suggestion
                    let is_message_complete (?) = self.incomplete_offset == self.buffers[self.active_buffer_index].len();
                    if  is_message_complete {
```
not sure about the naming, giving this comparison a name leads to better documented behavior

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+                    poll_reader(reader.as_raw_fd())?;
+                }
+                Err(error) => return Err(error),
+            }
+        }
+    }
+
+    /// Precondition: `read()` has been called once
+    pub fn get_last(&self) -> &StateStorage {
+        let inactive_buffer_index = (self.active_buffer_index + 1) % 2;
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:61]: This `[...] + 1 % 2` toggle computation is used multiple times? This could be a function? Or even more explicit, use an `enum` representing a tri-state instead of indices, which also covers the case of being uninitialized?

File: crates/hulk/src/nao/hula.rs [hunk 8]
```diff
@@ -363,7 +366,7 @@ impl From<types::Ear> for Ear {
     }
 }
 
-#[derive(Clone, Copy, Debug, Default)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
```
Comment (schmidma) [inline crates/hulk/src/nao/hula.rs:369]: why is this needed?

File: crates/hulk/src/nao/hula.rs [hunk 9]
```diff
@@ -377,10 +380,30 @@ pub struct ControlStorage {
...
+    stream: &mut UnixStream,
+    reader: &mut DoubleBufferedReader,
+) -> Result<StateStorage> {
+    reader
+        .read(stream, |file_descriptor| unsafe {
+            let mut set = MaybeUninit::<fd_set>::uninit();
+            FD_ZERO(set.as_mut_ptr());
+            let mut set = set.assume_init();
+            FD_SET(file_descriptor, &mut set);
+            if select(
```
Comment (schmidma) [inline crates/hulk/src/nao/hula.rs:393]: Can you elaborate what this does? Is there nothing that already implements such a behavior as an API function on some file descriptor type?

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+                    }
+                    poll_reader(reader.as_raw_fd())?;
+                }
+                Err(error) => return Err(error),
+            }
+        }
+    }
+
+    /// Precondition: `read()` has been called once
+    pub fn get_last(&self) -> &StateStorage {
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:60]: I'm not a fan of `get_[...]` maybe just drop the `get`-part?

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+    }
+}
+
+impl DoubleBufferedReader {
+    fn swap_buffers(&mut self) {
+        self.incomplete_offset = 0;
+        self.active_buffer_index = (self.active_buffer_index + 1) % 2;
+    }
+
+    pub fn read(
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:31]:
```suggestion
    pub fn draining_read(
```

??? What do you think?

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 10 above)
Comment (schmidma) [reply crates/hulk/src/nao/double_buffered_reader.rs:60]: Another suggestion is `buffered()`? Because `last` indicates, there is more to get, but the user can actually always only get one complete message out of the buffer. So there is no first or last, there is just one `buffered` message.

I'm open to for suggestions on the actual name, but i would like to see this concept represented.

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+    }
+
+    /// Precondition: `read()` has been called once
+    pub fn get_last(&self) -> &StateStorage {
+        let inactive_buffer_index = (self.active_buffer_index + 1) % 2;
+        unsafe { &*(self.buffers[inactive_buffer_index].as_ptr() as *const StateStorage) }
+    }
+}
+
+#[cfg(test)]
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:66]: I did not read the tests yet, but think about splitting the tests into "unit-tests" in this module file, and integration tests, which test the API available to the public user.

**Review by schmidma — CHANGES_REQUESTED** [review]
Some general remarks and mostly nit-picking on naming.

One general question:
You are concerned a lot about partial messages and stitching them back together afterwards. Is this actually happening? I assumed the unix socket to have extremely high memory bounds on the internal buffers? Or why are you concerned with partial messages? Please elaborate.

**Comment by h3ndrk** [comment]
> I assumed the unix socket to have extremely high memory bounds on the internal buffers?

I [read that Unix sockets have around 100-200 KB of buffer space](https://unix.stackexchange.com/a/424381). HULA sends messages of 652 bytes which means the buffer is full after around 300 messages which is approximately 3-4 seconds at a rate of 80 Hertz. When the buffer gets full, the likelihood that only a partial message still fits is very high. In my experiments, I saw around 80 KB of buffered messages to be usual during initialization. Hopefully, this explains the message stitching efforts.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 1 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs:9]: I wanted to define `pub struct DoubleBufferedReader<T>` but this does not work because I used `size_of::<T>()` in the type which is [currently unstable](https://github.com/rust-lang/rust/issues/60551). We could try but I thought, I don't need to spend effort generalizing for this single instantiation. But yes, if you have an idea, how we can implement it, I'd be happy to do so.

File: crates/hulk/src/nao/hula.rs (hunk 8 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/hula.rs:369]: Mainly to compare the types in assertions in the tests.

File: crates/hulk/src/nao/hula.rs (hunk 9 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/hula.rs:393]: Since this is Unix only functionality and Rust's standard library does not provide such high-level constructs, there is nothing available officially. I decided to use the libc call directly which is unsafe but has is a well-known API and does not introduce heavy abstractions. My goal was to find a thing that allows me to simply wait for a single file descriptor becoming readable. The other Linux polling methods would also work but `select()` is one of the simpler ways to do it. We could search for a safe abstraction over `select()` to get rid of the `unsafe {}` here. What is your opinion or do you know another way to use standard functionality that I didn't see?

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 12 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs:66]: IIRC testing public API is not the separation argument between unit and integration tests. Integration tests are testing a unit together with other units. Since I used traits to only implement the functionality that I need, I would argue that the `DoubleBufferedReader` is a unit and all tests in this section mock all I/O and therefore fall into the unit-test category. Correct me if I'm wrong. :wink:

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 2 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: It is now named `number_of_read_bytes_in_active_buffer`.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 3 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: Abstracted.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 4 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: The reader and poller are now wrapped in the `DoubleBufferedReader`.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 5 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: It is now named `is_at_least_one_buffer_complete`.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 6 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: Is extracted as `is_active_buffer_complete` to align with the `is_at_least_one_buffer_complete` variable.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 7 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: The index calculations are now extracted as methods.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 10 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: The `DoubleBufferingReader::drain()` now returns the read value directly. The precondition is also enforced by that design. `comments -= 1`

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 11 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: I called it not `drain()`. I'm not 100% happy with this name because it actually drains and returns the last complete value. `drain()` does not reflect the value returning but maybe that is okay.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 12 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs:113]: We could also add integration tests but I think this piece of code will be tested enough in using the NAO in everyday life situations.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 1 above)
Comment (schmidma) [reply crates/hulk/src/nao/double_buffered_reader.rs]: Hmm.. i'm fine with either approach. One solution might be to drop `size_of::<T>()` and replace it with `const N: usize` or similar.
Another idea, can you just make an array of `T` and reinterpret it to a `[u8]` when reading from the socket?

File: crates/hulk/src/nao/hula.rs (hunk 9 above)
Comment (schmidma) [reply crates/hulk/src/nao/hula.rs]: No, I just assumed, there already is a safe abstraction. I now, that for example https://docs.rs/tokio/latest/tokio/io/unix/struct.AsyncFd.html#method.readable does something like you did manually. Maybe there is a crate for these functionalities out there?!

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 11 above)
Comment (schmidma) [reply crates/hulk/src/nao/double_buffered_reader.rs]: but `draining_read` does? What is your objection to that name?

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 12 above)
Comment (schmidma) [reply crates/hulk/src/nao/double_buffered_reader.rs:113]: Maybe my current neovim setup just lacks easy folding of parts, but I am always a little distracted by these huge parts of testing code in the implementation. But I think there is not a pretty fix to it. Feel free to resolve when read or start a discussion ;)

**Comment by schmidma** [comment]
But did you ever observe partial messages in the socket buffer? And what is HULA doing then? Is it blocking? Crashing?!

File: crates/hulk/src/nao/hula.rs (hunk 8 above)
Comment (schmidma) [reply crates/hulk/src/nao/hula.rs:367]: Ok, I was shortly thinking about a way to represent the fact, that it is only needed for testing. But never mind. I'm fine, if you have no further ideas.
(Feel free to resolve when read)

File: crates/hulk/src/nao/double_buffered_reader.rs [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+
+use libc::{fd_set, select, FD_SET, FD_ZERO};
+
+use super::hula::StateStorage;
+
+const NUMBER_OF_BUFFERS: usize = 2;
+
+pub struct DoubleBufferedReader<Reader, Poller> {
+    reader: Reader,
+    poller: Poller,
```
Comment (schmidma) [inline crates/hulk/src/nao/double_buffered_reader.rs:16]: Why is it generic over the `poller`?

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 7 above)
Comment (schmidma) [reply crates/hulk/src/nao/double_buffered_reader.rs]: What do you think about the most explicit approach with `enum`s? I'm not a fan of indices

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 12 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs:113]: Well we could move the code into its own file but then we would do it differently than in the other files. Do we want this Präzedenzfall? Do we want this everywhere? I think it is okay to have it near the implementations since it is always at the bottom and noone forces you to scroll that far. Now you resolve again 😉

File: crates/hulk/src/nao/hula.rs (hunk 9 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/hula.rs]: Maybe this? https://github.com/DoumanAsh/selecting Sadly there is no `readable()` in non-async land as far as I know. IMHO the benefit for such simple functionality is debatable.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 7 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: At some point we need the indices again if we want to peek into the `Self::buffers`. IMHO it is okay to use indices consistently and as a bonus we are general about the number of buffers.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 13 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs:16]: For dependency injection for testability.

**Comment by h3ndrk** [comment]
> But did you ever observe partial messages in the socket buffer?

No, the scenario would happen if HULK pauses reading for 3-4 seconds, then reads faster than HULA being able to write the remaining message. This could happen if the scheduler decides to not schedule HULA for a short moment in time.

> And what is HULA doing then? Is it blocking?

Yes

> Crashing?!

No, but it will wait for everything to be written until reading more messages from LoLA.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 1 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: It is now generic over the read type. It is still located in the `hulk` crate because this is the single usage for now. We can extract it into a crate later.

File: crates/hulk/src/nao/double_buffered_reader.rs (hunk 11 above)
Comment (h3ndrk) [reply crates/hulk/src/nao/double_buffered_reader.rs]: Renamed to `draining_read()`.

File: crates/hulk/src/nao/hula.rs [hunk 14]
```diff
@@ -377,10 +378,16 @@ pub struct ControlStorage {
     pub stiffness: JointsArray,
 }
 
-pub fn read_from_hula(stream: &mut UnixStream) -> Result<StateStorage> {
-    let mut read_buffer = [0; size_of::<StateStorage>()];
-    stream.read_exact(&mut read_buffer)?;
-    Ok(unsafe { read(read_buffer.as_ptr() as *const StateStorage) })
+pub fn read_from_hula<Reader, Poller>(
```
Comment (schmidma) [inline crates/hulk/src/nao/hula.rs:381]: Does this need to be generic over the `Poller`?

Comment (h3ndrk) [reply crates/hulk/src/nao/hula.rs]: Removed generic parameters of reader and poller.

**Comment by h3ndrk** [comment]
@schmidma Tested successfully on a NAO with legacy HULA.

**Review by schmidma — APPROVED** [review]
